_posts/2019-02-25-practice-oop-to-front-end-universal-state-module.md

```diff
@@ -0,0 +1,291 @@
+
+The following is the connector section using `usm-redux`:
+
+```js
+// index.js
+export const counter = Counter.create();
+
+ReactDOM.render(
+  <Provider store={counter.store}>
+    <App />
+  </Provider>,
+  document.getElementById('root')
+);
+```
+
+```js
+// app.js
+import { connect } from 'react-redux';
+import { counter } from './';
+
+export default connect(
+  state => ({ ount: state.count })
+)( props => 
+  <div>
+    <button onClick={() => counter.increase()}>+</button>
+    {props.count}
+    <button onClick={() => counter.decrease()}>-</button>
+  </div>
+);
+```
+
+And here is the same counter `View` connector using `usm-mobx`:
+
+```js
+// index.js
+
+export const counter = Counter.create();
+
+ReactDOM.render(
+  <App />,
+  document.getElementById('root')
+);
+```
+
+
+```js
+// app.js
+import { observer } from 'mobx-react';
+import { counter } from './';
+
+export default observer(() =>
+  <div>
+    <button onClick={() => counter.increase()}>+</button>
+    {counter.count}
+    <button onClick={() => counter.decrease()}>-</button>
+  </div>
+);
+```
+
+In the `usm-redux` and `usm-mobx` examples, in addition to some differences in the View connector, the state module part is in accord, and this is the universal state module that we had proposed.
+
+** USM currently supports Redux, MobX, Vuex and Angular.**
+
+### Features
+
+- Universal State Management
+- Standardized Module Lifecycle
+- Optional Event System
+- Support Stateless Model
+- Support React/Vue/Angular
+
+### Decorators
+
+`usm` provides decorator `@state` to wrap a variable with a state, and decorator `@action` is used to wrap a function that changes state (the last parameter passed in by the function is always the current state object).
+
+```js
+class Shop extends Module {
+  @state goods = [];
+  @state status = 'close';
+
+  @action
+  operate(item, status, state) {
+    state.goods.push(item);
+    state.status = status;
+  }
+  // call function -> this.operate({ name: 'fruits', amount: 10 }, 'open');
+}
+```
+
+### Module lifecycle
+
+`usm` provides these lifecycle events:
+
+- `moduleWillInitialize`
+- `moduleWillInitializeSuccess`
+- `moduleDidInitialize`
+- `moduleWillReset`
+- `moduleDidReset`
+
+The order in which they are run is shown in the following flow chart:
+
+![lifecycle](/integration-blog/assets/2019-02-25-practice-oop-to-front-end-universal-state-module/usm_lifecycle.png)
+
+These module lifecycles can be used to coordinate module initialization dependencies.
+
+```js
+class TodoList extends Module {  
+  async moduleWillInitialize() {
+    console.log(
+      'TodoList -> moduleWillInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleWillInitializeSuccess() {
+    console.log(
+      'TodoList -> moduleWillInitializeSuccess',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
```

Comment:
pendding => pending
...
You obviously copy pasted these console.log statements...